### ISSUE TITLE
Adding non-lazy functions that parse R6 query methods

### ISSUE THREAD

**Issue by tbradley1013** [rstudio/connectapi#34]
Hi all, 

Is there any plan to expose the R6 query methods a little more and/or provide functions that will parse the output of these methods? Providing these query/parse functions may be a alternative while the full implementation of the lazy query approaches (e.g. filter methods) are still being developed. 

Here is an example for the shiny usage method, and it shouldn't be hard to expand to the other methds used in `tbl_connect` too. I simply took your code from `api_build.op_base_connect` and exposed the rest of the API request parameters:

``` r
library(connectapi)

client <- connect()
#> Defining Connect with host: 


get_usage <- function(src, content_guid = NULL, 
                      min_data_version = NULL,
                      from = NULL,
                      to = NULL,
                      limit = 20,
                      previous = NULL,
                      nxt = NULL,
                      asc_order = TRUE){
  connectapi:::validate_R6_class("Connect", src)
  
  res <- src$inst_shiny_usage(
    content_guid = content_guid, 
    min_data_version = min_data_version, 
    from = from,
    to = to, 
    limit = limit, 
    previous = previous, 
    nxt = nxt,
    asc_order = asc_order
  )
  
  res <- page_cursor(src, res, limit = limit)
  
  out <- purrr::map_df(
    res, 
    function(x) {
      purrr::map(
        .x = x,
        .f = function(y) {
          prep <- purrr::pluck(y, .default = NA)
          # TODO: Should figure out what we want to do about sub-objects...
          # i.e. content: git details... could build a nested list...?
          if (length(prep) > 1)
            prep <- NA
          return(prep)
        }
      )
    }
  )
  
  return(out)
}


get_usage(src = client, from = as.POSIXct("2019-11-23"), to = as.POSIXct("2019-11-25"))
#> # A tibble: 1 x 5
#>   content_guid             user_guid started       ended       data_version
#>   <chr>                    <lgl>     <chr>         <chr>              <int>
#> 1 xxxxxxxx-xxxx-xxxx-xxxx~ NA        2019-11-23T1~ 2019-11-23~            1
```

<sup>Created on 2019-11-26 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

As for pagination in this example, you could add a filter at the end that just selects only the top n results based on what limit was.

**Comment by tbradley1013** [comment]
I am happy to put them together and create a PR if you are interested and think it is worthwhile

**Comment by colearendt** [comment]
I think this is a great idea, and should be pretty easy to implement! I'm definitely amenable to a PR if you're interested in taking a stab at it. 😄

**Comment by tbradley1013** [comment]
Sounds good. Any preference on the function names? Would you prefer them to be the same as the R6 method names or maybe something like my example of `get_usage` or maybe `get_shiny_usage`

**Comment by colearendt** [comment]
Naming is so hard 🙈I like the [tidyverse conventions](https://style.tidyverse.org/functions.html#naming) but find it a bit hard to apply in cases like this. Is `get_inst_shiny_usage` too long? I like `get_shiny_usage`, but imagine it would be nice to give a common prefix to all of the "instrumentation" endpoints, since our aim is that there will be more as time goes on 😄

**Comment by tbradley1013** [comment]
Personally, I think it is a little long, but if there are going to be a lot of additional endpoints in the instrumentation group then it may be the way to go. My only thought is that people who are just using the package and are not as familiar with the API endpoints themselves may be confused by the inst portion.... 

We could potentially export the `inst` functions as both `get_inst_shiny_usage` and `get_shiny_usage` and deprecate one later after user feedback? I am not sure whether that is a sustainable approach though. 

Also, on a side note, I have a potential fix to the pagination thing that I will submit as a separate PR

**Comment by slopp** [comment]
Inst doesn’t mean instrumentation to me, but if you want a common prefix you could do:

get_usage_shiny
get_usage_static

**Comment by colearendt** [comment]
I like that! And great points @tbradley1013 ! Definitely please feel free - PRs are welcome! 😄

**Comment by tbradley1013** [comment]
Sounds good. Do you know if there is documentation for the `applications` endpoint that is used by the `get_apps` R6 method? I am not seeing it in the [API Reference](https://docs.rstudio.com/connect/api/#getContent) and would like to be able to add some detailed about the options that can be provided to the `filter` argument

**Comment by colearendt** [comment]
Yeaaah. `get_apps` is not a public API 😅 (i.e. so we _really_ shouldn't be using it). So I'm not sure how much we want to build into that one. If/when it becomes a public API, it would likely change a good bit, so probably worth punting for now. Or just taking a stab at what you might find useful and then calling it good. It's probably not worth the effort to shoot for "feature completeness."

**Comment by tbradley1013** [comment]
Interesting. Well I may play around with it a little but for now, i will just expose the filter argument and let users try to play with it too. 

Is there a place to submit the feedback you are referring to about the API and what is useful? I saw a comment about the experimental endpoints in the reference guide that mentioned their finality would be based on user feedback but it didn't mention where to send that feedback. There are a few additional things I would like added if possible!

**Comment by colearendt** [comment]
That is an amazing point... I have no idea where to send that feedback haha. If you want to dump it here in an issue, I can pass it along for now 😅 I think https://community.rstudio.com/c/r-admin is probably the current "home" for feedback if you want to dump into a new thread there too. What's great about that forum is that we can ask for clarifying questions and link those threads for developers as well.

**Comment by tbradley1013** [comment]
Ah yes, I should have thought of doing it on the community.

**Comment by slopp** [comment]
@tbradley1013 I'd be happy to chat with you as well if you just want to dump thoughts on the app listing API and what you're hoping to do with this package more generally, and could probably get a few engineers to join as well.

**Comment by tbradley1013** [comment]
@slopp that sounds good. Let me know how you want to facilitate that (i.e. email, community chat, or phone call). I am out of the office until Monday but am happy to chat sometime next week if you prefer phone.

**Comment by slopp** [comment]
Let's setup time next week, can you shoot me an email at sean@rstudio.com ?  Thanks again!

**Comment by colearendt** [comment]
I believe this was closed by #36? Please feel free to reopen if not!
